src/t_hash.c

```diff
@@ -922,3 +943,213 @@ void hscanCommand(client *c) {
+     * HRANDFIELD_SUB_STRATEGY_MUL times the number of requested elements.
+     * In this case we create a hash from scratch with all the elements, and
+     * subtract random elements to reach the requested number of elements.
+     *
+     * This is done because if the number of requested elements is just
+     * a bit less than the number of elements in the hash, the natural approach
+     * used into CASE 4 is highly inefficient. */
+    if(count*HRANDFIELD_SUB_STRATEGY_MUL > size) {
+        /* Add all the elements into the temporary dictionary. */
+        while((hashTypeNext(hi)) != C_ERR) {
+            int ret = DICT_ERR;
+            sds key, value = NULL;
+
+            key = hashTypeCurrentObjectNewSds(hi,OBJ_HASH_KEY);
+            if (withvalues)
+                value = hashTypeCurrentObjectNewSds(hi,OBJ_HASH_VALUE);
+            ret = dictAdd(d, key, value);
+
+            serverAssert(ret == DICT_OK);
+        }
+        serverAssert(dictSize(d) == size);
+
+        /* Remove random elements to reach the right count. */
+        while(size > count) {
+            dictEntry *de;
+            de = dictGetRandomKey(d);
+            dictUnlink(d,dictGetKey(de));
+            sdsfree(dictGetKey(de));
+            sdsfree(dictGetVal(de));
+            dictFreeUnlinkedEntry(d,de);
+            size--;
+        }
+    }
+
+    /* CASE 4: We have a big hash compared to the requested number of elements.
+     * In this case we can simply get random elements from the hash and add
+     * to the temporary hash, trying to eventually get enough unique elements
+     * to reach the specified count. */
+    else {
+        unsigned long added = 0;
+        sds key, value = NULL;
+        while(added < count) {
+            hashTypeRandomElement(hash, size, &key, withvalues? &value : NULL);
+
+            /* Try to add the object to the dictionary. If it already exists
+            * free it, otherwise increment the number of objects we have
+            * in the result dictionary. */
+            if (dictAdd(d,key,value) == DICT_OK){
+                added++;
+            } else {
+                sdsfree(key);
+                sdsfree(value);
+            }
+        }
+    }
+
+    {
+        if (withvalues && c->resp == 2)
+            addReplyArrayLen(c, dictSize(d)*2);
+        else
+            addReplyArrayLen(c, dictSize(d));
+        dictIterator *di;
+        dictEntry *de;
+        di = dictGetIterator(d);
+        while ((de = dictNext(di)) != NULL) {
+            sds key = dictGetKey(de);
+            sds value = dictGetVal(de);
+            if (withvalues && c->resp > 2)
+                addReplyArrayLen(c,2);
+            addReplyBulkSds(c, key);
+            if (withvalues)
+                addReplyBulkSds(c, value);
+        }
+
+        hashTypeReleaseIterator(hi);
+        dictReleaseIterator(di);
+        dictRelease(d);
+    }
+}
+
+/* HRANDFIELD [<count> WITHVALUES] */
```

Comment:
From the original request, I would guess that the author would be using HRANDFIELD 1 WITHVALUES, since they want to process the record they pull. So unlike SRANDMEMBER, where count is unlikely to be 1, I think it's very likely someone will call HRANDFIELD with count of 1. So, I think the optimized n=1 case should also be handled hrandfieldWithCountCommand and the code always flows into that function.

Reply:
i'll do some optimization to CASE4 for that.